### ISSUE TITLE
Add Soft-DTW distance and alignment

### ISSUE THREAD

**Issue by sipemu** [sipemu/fdars#11]
## Summary

Implement **Soft-DTW**, a differentiable relaxation of Dynamic Time Warping, as a distance metric and alignment method.

## Motivation

Standard DTW (already available via `metric.DTW()`) uses a hard minimum in the recursive cost computation, making it non-differentiable. **Soft-DTW** replaces the hard minimum with a soft minimum (log-sum-exp), producing a **differentiable** distance that:

1. Can be used as a loss function in gradient-based optimization and neural networks
2. Provides smoother distance surfaces (better behaved for clustering/MDS)
3. Interpolates between DTW ($\gamma \to 0$) and Euclidean distance ($\gamma \to \infty$) via a smoothing parameter $\gamma$

$$\text{sdtw}_\gamma(x, y) = \min^\gamma_{A \in \mathcal{A}} \langle A, \Delta(x, y) \rangle$$

where $\min^\gamma$ is the soft-minimum operator: $\min^\gamma(a_1, \ldots, a_n) = -\gamma \log \sum_i e^{-a_i/\gamma}$.

## Proposed API

```r
# Soft-DTW distance matrix
D <- metric.softDTW(fdataobj, gamma = 1.0)

# Available through the metric dispatcher
D <- metric(fdataobj, method = "softDTW", gamma = 1.0)

# Soft-DTW barycenter (mean under soft-DTW geometry)
bc <- softdtw.barycenter(fdataobj, gamma = 1.0, max.iter = 50)
```

## Key features

- Smoothing parameter $\gamma$ controls the softness (DTW ↔ Euclidean interpolation)
- **Soft-DTW barycenter**: a principled mean shape under soft-DTW, analogous to the Karcher mean for elastic
- Rust backend for performance (the DP recursion is the same structure as DTW)
- Normalized variant (soft-DTW divergence) that satisfies $\text{sdtw}_\gamma(x, x) = 0$

## Use cases

- Distance metric for clustering (`cluster.kmeans`, `cluster.fcm`) as an alternative to elastic/DTW
- Time series classification with gradient-based methods
- More robust than hard DTW for noisy signals

## References

- Cuturi, M. and Blondel, M. (2017). Soft-DTW: A differentiable loss function for time-series. *Proceedings of the 34th International Conference on Machine Learning (ICML)*, 894–903.
- Blondel, M., Mensch, A., and Vert, J.-P. (2021). Differentiable divergences between time series. *Proceedings of the 24th International Conference on Artificial Intelligence and Statistics (AISTATS)*.
